tests/src/Kernel/Entity/OgMembershipTest.php

```diff
@@ -489,6 +489,21 @@ public function testMemberRole() {
     $this->assertEquals(OgRoleInterface::AUTHENTICATED, $role->getName());
   }
 
+  /**
+   * Tests that we can retrieve the (empty) roles list from a new membership.
+   *
+   * If a membership is newly created and doesn't have a group associated with
```

Comment:
I think we should prohibit this case if possible. That is, upon create - already pass the group and user.

Reply:
I'm doing this on an empty entity because it is the simplest way to prove the fatal error. In real life this occurs because the group entity is asynchronously deleted and memberships that are statically cached still reference it. This is not something we can prevent by simply requiring a group to be passed on instantiation. If you look at `OgMembership::getGroup()`, it gets the group entity from the entity storage, and this can always be deleted outside of the control of the membership.

```
// Create membership with a valid group.
$group = EntityTest::create();
$membership = Og::createMembership($group, $this->user);
$membership->save();

// Delete the group.
$group->delete();

// Fatal error.
$membership->getGroup()->bundle();
```

I had a look if it was feasible to prevent to create memberships without groups. I'm afraid it would have negative consequences because we would essentially be throwing an exception in `EntityStorageBase::create()` during the `::preCreate()` phase. It is legal to call `EntityStorageBase::create()` without passing any `$values`, so it is likely that there is a lot of code out there in the wild that does this.

I was looking for some examples in core and contrib where this pattern is used and it seems we are even doing it ourselves! In `MembershipManager::createMembership()` we create an empty entity, populate it and then save it. It looks like pretty standard Drupal code, I can imagine this type of code being very widespread:

```
  public function createMembership(EntityInterface $group, AccountInterface $user, $membership_type = OgMembershipInterface::TYPE_DEFAULT) {
    $membership = OgMembership::create(['type' => $membership_type]);
    $membership
      ->setOwner($user)
      ->setGroup($group);
    return $membership;
  }
```

So I had a further look, to see if the memberships have adequate protection for making sure they have correct groups and roles associated with it. It seems pretty good. We are validating them on presave, so we are basically guaranteeing that when the memberships are saved they have correct references.

Reply:
Understood. Thanks